storage/innobase/trx/trx0trx.cc

```diff
@@ -1501,7 +1505,9 @@ TRANSACTIONAL_INLINE inline void trx_t::commit_in_memory(const mtr_t *mtr)
     trx_finalize_for_fts(this, undo_no != 0);
 
 #ifdef WITH_WSREP
-  ut_ad(is_wsrep() == wsrep_on(mysql_thd));
+  if (!mysql_thd || !wsrep_get_assert(mysql_thd, WSREP_ASSERT_INNODB_TRX)) {
```

Comment:
Having a function in wsrep service just for avoiding this assert seems a bit excessive to me. Would there be some more lightweight solution for this, like 

```
/* Assert that any transaction which is started as a wsrep transaction, also commits
* or rolls back as a wsrep transaction. BF transactions may turn wsrep off temporarily
* for rollback when they are retried. */
ut_ad(is_wsrep() == wsrep_on(mysql_thd) || (in_rollback && wsrep_thd_is_BF(mysql_thd));
```

Reply:
I agree Teemu and comment above makes things a lot clearer.

Reply:
Yes, that is a more light-weight solution. 
I suggested this new function in anticipation that there would be similar problems in the future:
this new function would allow solving them easily.

Your suggestion is certainly simpler and nicer if we think that this problem is just one of a kind.

Reply:
Re-implemented the fix:

    Modified the bug fix for MDEV-36554 by replacing the general
    mechanism for disabling assertions in InnoDB code with solution
    specific to this rollback issue.
    
    The appliers sets a flag in the applier thread for the duration of the
    local rollback. This is used in InnoDB code to detect a local
    rollback.